src/util.cc

```diff
@@ -622,3 +624,44 @@ bool Truncate(const string& path, size_t size, string* err) {
   }
   return true;
 }
+
+bool ReplaceContent(const string& file_dst, const string& new_content, string* err) {
+#ifdef _WIN32
+  if (!DeleteFile(file_dst.c_str())) {
+    *err = GetLastErrorString();
+    return false;
+  }
+
+  if (rename(new_content.c_str(), file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+#else
+  struct stat old_file;
+  bool found_uid_gid = true;
+
+  if (stat(file_dst.c_str(), &old_file) < 0) {
+    found_uid_gid = false;
+  }
+
+  if (unlink(file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  if (rename(new_content.c_str(), file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+
+  // apply uid and gid again so we always stay as the uid and gid of the first caller that created a file
```

Comment:
What do you mean by "first caller"?

Reply:
first caller of ninja

Reply:
Maybe saying "ninja invocation" instead of "caller" would be clearer wording.